### DISCUSSION TITLE
Device Twin Message Confirmation

### DISCUSSION THREAD

**Discussion by MMaiero** [Azure/azure-iot-sdk-java#1471] [Ideas]
- **OS and version used:** Raspbian Buster

- **Java runtime used:** openjdk version "1.8.0_181"

- **SDK version used:** 0.22.0
iot-device-client version 1.16.0

# Description of the issue:
In our context, we would like to get a confirmation that the Device Twin message has been received by the cloud platform, just like for the plain telemetry messages. I see that the APIs provide a way to get notification of the property version, but I am not able to get a notification that the whole massage has been received. Is it possible to achieve that result with the current API?

# Code sample exhibiting the issue:

# Console log of the issue:

# Need Support?
* Have a feature request for SDKs? Please post it on [User Voice](https://feedback.azure.com/forums/321918-azure-iot) to help us prioritize.
* Have a technical question? Ask on [Stack Overflow](https://stackoverflow.com/questions/tagged/azure-iot-hub) with tag “azure-iot-hub”
* Need Support? Every customer with an active Azure subscription has access to support with guaranteed response time.  Consider submitting a ticket and get assistance from Microsoft support team
* Found a bug? Please help us fix it by thoroughly documenting it and filing an issue on GitHub (C, Java, .NET, Node.js, Python).

**Comment by timtay-microsoft** [comment]
> we would like to get a confirmation that the Device Twin message has been received by the cloud platform

Which twin message are you referring to here? A reported property update message?

**Comment by MMaiero** [comment]
Oh, sorry. I was referring to the reported property messages.

**Comment by timtay-microsoft** [comment]
Okay, just wanted to double check. Thanks for the clarification.

A good example of using device twin can be found in [this sample](https://github.com/Azure/azure-iot-sdk-java/blob/master/device/iot-device-samples/device-twin-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/DeviceTwinSample.java).

In it, the device client provides a callback (deviceTwinStatusCallback) to the startDeviceTwin(...) call that will be fired each time a twin message was successfully sent, including reported property messages. Note that it does also fire when the first twin subscription message is sent as well.

**Comment by timtay-microsoft** [comment]
@MMaiero did you have any other questions on this topic? If not, I will close this issue

**Comment by MMaiero** [comment]
Hi @timtay-microsoft,
yes I've already used that example as reference. What is not clear to me is how to know if a certain property is actually acknowledged by the cloud. I mean: I receive the reported property in the onProperty method and I know that it has version 50. But I do I relate that version with the original message sent from the device?

Don't know if it is clear what I mean

**Comment by timtay-microsoft** [comment]
When you receive the onProperty callback, that represents IotHub accepting your reported property update. You don't really need to check the version here, or assign any version when sending subsequent reported property updates. That version is just a counter maintained on the hub side that tracks how many updates the twin has had.

Does that make sense?

**Comment by MMaiero** [comment]
If I'm sending twin updates as reported properties one after the other, maybe in a problematic environment like using a 3G connection, how do I know that all my updates got received from the cloud. This is actually possible with telemetry message and not really for twin properties. Isn't that a little bit asymmetric?

Thanks

**Comment by timtay-microsoft** [comment]
The reported properties callback should be fired each time after the IoT Hub acknowledges receiving a reported properties update. You will know if your hub never received your reported property update if you never receive the callback corresponding to that property update.

**Comment by MMaiero** [comment]
Hi @timtay-microsoft thank you for your response.
Maybe I'm doing something wrong, but the only callback that is invoked in my case, on reported properties update, is the DeviceTwinStatusCallBack that reports only the status. I was expecting to get a TwinPropertyCallBack but from [here](https://github.com/Azure/azure-iot-sdk-java/blob/master/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceTwin/DeviceTwin.java#L123) I see that the twinPropertycallback is not invoked in case of DEVICE_OPERATION_TWIN_UPDATE_REPORTED_PROPERTIES_RESPONSE.

Am I in the right path?

Thanks

**Comment by MMaiero** [comment]
By the way, I have tried to mockup what I originally intended, just to give you an idea.
I have committed it here: https://github.com/MMaiero/azure-iot-sdk-java/tree/twin_callback

Thanks

**Comment by timtay-microsoft** [comment]
After taking another look at this issue, here is where the SDK stands right now. For sending a reported property, there is no way for a user to know when that reported property was acknowledged by the service. I consider this a gap in our SDK, as you are right to expect some way to know when these messages get ack'd. This may take a bit of an overhaul to fix without breaking existing users, but we'll look into it

**Comment by MMaiero** [comment]
Hi @timtay-microsoft, first of all thank you for the update. 
I'm not aware of the latest changes but the proposed update in the comment above is not working in your opinion? And I believe it's not a breaking change for existing APIs, right?

Thanks again
